keanu-project/src/test/java/io/improbable/keanu/algorithms/sampling/ForwardSamplerTest.java

```diff
@@ -0,0 +1,233 @@
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.binary.DoubleBinaryOpVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.ProbabilisticDouble;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+
+@Slf4j
+public class ForwardSamplerTest {
+
+    private KeanuRandom random;
+
+    @Before
+    public void setup() {
+        random = new KeanuRandom(1);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwIfSampleFromHasObservationsDownstreamOfLatents() {
+        GaussianVertex A = new GaussianVertex(100.0, 1);
+        DoubleVertex B = A.plus(ConstantVertex.of(5.));
+        GaussianVertex C = new GaussianVertex(B, 1.);
+        C.observe(100.);
+
+        BayesianNetwork network = new BayesianNetwork(A.getConnectedGraph());
+        ProbabilisticModel model = new KeanuProbabilisticModel(network);
+
+        Keanu.Sampling.Forward.withDefaultConfig(random).getPosteriorSamples(model, Arrays.asList(A, B), 1000);
+    }
+
+    @Test
+    public void canSampleFromPrior() {
+        GaussianVertex A = new GaussianVertex(100.0, 1);
+        DoubleVertex B = A.plus(ConstantVertex.of(5.));
+
+        BayesianNetwork network = new BayesianNetwork(A.getConnectedGraph());
+        ProbabilisticModel model = new KeanuProbabilisticModel(network);
+
+        final int sampleCount = 1000;
+        NetworkSamples samples = Keanu.Sampling.Forward.withDefaultConfig(random).getPosteriorSamples(model, Arrays.asList(A, B), sampleCount);
+
+        double averageA = samples.getDoubleTensorSamples(A).getAverages().scalar();
+        double averageB = samples.getDoubleTensorSamples(B).getAverages().scalar();
+
+        assertEquals(sampleCount, samples.size());
+        assertEquals(100.0, averageA, 0.1);
+        assertEquals(105.0, averageB, 0.1);
+    }
+
+    @Test
+    public void nonProbabilisticVerticesAreRecomputedDuringForwardSample() {
+        AtomicInteger opCount = new AtomicInteger(0);
+
+        GaussianVertex A = new GaussianVertex(0, 1);
+        TestGraphGenerator.PassThroughVertex B = new TestGraphGenerator.PassThroughVertex(A, opCount, null, id -> log.info("OP on id:" + id));
+
+        BayesianNetwork network = new BayesianNetwork(A.getConnectedGraph());
+        ProbabilisticModel model = new KeanuProbabilisticModel(network);
+
+        Keanu.Sampling.Forward.withDefaultConfig(random).getPosteriorSamples(model, Arrays.asList(A, B), 100);
+
+        assertEquals(100, opCount.get());
+    }
```

Comment:
suggestion - How to do this with mocking instead of the `PassThroughVertex`:

    @Test
    public void nonProbabilisticVerticesAreRecomputedDuringForwardSample() {
        GaussianVertex A = new GaussianVertex(0, 1);
        Vertex B = mock(Vertex.class);
        when(B.getParents()).thenReturn(Collections.singleton(A));
        when(B.getId()).thenReturn(new VertexId());

        ProbabilisticModel model = new KeanuProbabilisticModel(A.getConnectedGraph());

        Keanu.Sampling.Forward.withDefaultConfig(random).getPosteriorSamples(model, Arrays.asList(A, B), 100);

        verify(B, times(100)).setValue(any());
    }

Reply:
Nice - done